modules/gdnative/videodecoder/video_stream_gdnative.cpp

```diff
@@ -0,0 +1,336 @@
+				if (pos < 0 && -pos > file->get_position()) {
+					return -1;
+				}
+				pos = pos + static_cast<int>(file->get_position());
+				file->seek(pos);
+				pos = static_cast<int64_t>(file->get_position());
+				return pos;
+			} break;
+			case SEEK_END: {
+				// Just in case something goes wrong
+				if (-pos > len) {
+					return -1;
+				}
+				file->seek_end(pos);
+				pos = static_cast<int64_t>(file->get_position());
+				return pos;
+			} break;
+			default: {
+				// Only 4 possible options, hence default = AVSEEK_SIZE
+				// Asks to return the length of file
+				return static_cast<int64_t>(len);
+			} break;
+		}
+	}
+	// In case nothing works out.
+	return -1;
+}
+
+void GDAPI godot_videodecoder_register_decoder(const godot_videodecoder_interface_gdnative *p_interface) {
+	printf("REGD2\n");
+	decoder_server.register_decoder_interface(p_interface);
+}
+}
+
+// VideoStreamPlaybackGDNative starts here.
+
+bool VideoStreamPlaybackGDNative::open_file(const String &p_file) {
+	ERR_FAIL_COND_V(interface == NULL, false);
+	file = FileAccess::open(p_file, FileAccess::READ);
+	bool file_opened = interface->open_file(data_struct, file);
+
+	num_channels = interface->get_channels(data_struct);
+	mix_rate = interface->get_mix_rate(data_struct);
+
+	godot_vector2 vec = interface->get_texture_size(data_struct);
+	texture_size = *(Vector2 *)&vec;
+
+	pcm = (float *)memalloc(num_channels * AUX_BUFFER_SIZE * sizeof(float));
+	memset(pcm, 0, num_channels * AUX_BUFFER_SIZE * sizeof(float));
+	pcm_write_idx = -1;
+	samples_decoded = 0;
+
+	texture->create((int)texture_size.width, (int)texture_size.height, Image::FORMAT_RGBA8, Texture::FLAG_FILTER | Texture::FLAG_VIDEO_SURFACE);
+
+	return file_opened;
+}
+
+void VideoStreamPlaybackGDNative::update(float p_delta) {
+	if (!playing || paused) {
+		return;
+	}
+	if (!file) {
+		return;
+	}
+	time += p_delta;
+	ERR_FAIL_COND(interface == NULL);
+	interface->update(data_struct, p_delta);
+
+	if (pcm_write_idx >= 0) {
+		// Previous remains
+		int mixed = mix_callback(mix_udata, pcm, samples_decoded);
+		if (mixed == samples_decoded) {
+			pcm_write_idx = -1;
+		} else {
+			samples_decoded -= mixed;
+			pcm_write_idx += mixed;
+		}
+	}
+	if (pcm_write_idx < 0) {
+		samples_decoded = interface->get_audioframe(data_struct, pcm, AUX_BUFFER_SIZE);
+		pcm_write_idx = mix_callback(mix_udata, pcm, samples_decoded);
+		if (pcm_write_idx == samples_decoded) {
+			pcm_write_idx = -1;
+		} else {
+			samples_decoded -= pcm_write_idx;
+		}
+	}
+
+	// printf("R: %i\tM %i\n", samples_decoded, pcm_write_idx);
+
+	while (interface->get_playback_position(data_struct) < time) {
+
+		update_texture();
+	}
+}
+
+void VideoStreamPlaybackGDNative::update_texture() {
+	PoolByteArray *pba = (PoolByteArray *)interface->get_videoframe(data_struct);
+
+	if (pba == NULL) {
+		playing = false;
+		return;
+	}
+
+	Ref<Image> img = memnew(Image(texture_size.width, texture_size.height, 0, Image::FORMAT_RGBA8, *pba));
```

Comment:
Doable. I had merely used what webm seems to use.

Reply:
I'm confused as to finding a way to not repeat allocation. Because even set_data from buffer in Image class is essentially allocating Image, and copying data from the new Image.

Reply:
Image::FORMAT_RGBA8 sucks :( it would be much more efficient to allow the plugin to choose the format